### PR TITLE
llvmPackages: llvmPackages_{16,18} -> llvmPackages_19

### PR DESCRIPTION
**Pull request by emilazy**
This is the work‐in‐progress PR for the 25.05 bump to LLVM 19. We’ve had a few Darwin people working on this independently, but the issues and patches are piling up and it’d be good to have a common branch to work on together. It’s currently based on my [Darwin minimum version bump PR](https://github.com/NixOS/nixpkgs/pull/352129) since that will land in the same cycle, but will be retargeted to `staging` once the branch‐off happens and that PR is merged. Please feel free to send PRs to this branch.

Known blockers:

* [x] https://github.com/NixOS/nixpkgs/issues/352724 – worked around, still not understood
* [x] Cursed Rust `wasm32-unknown-unknown` stuff that @paparodeo is working on – still cursed, but clean‐ups have made it broken in a less visible way

cc @NixOS/llvm @NixOS/darwin-core

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by paparodeo — APPROVED** [review]

**Review by RossComputerGuy — APPROVED** [review]

**Comment by emilazy** [comment]
It turns out that LLVM 19 defines `TARGET_OS_*` macros itself, presumably like the Xcode toolchain has already been doing, so we can get rid of a bunch of hacks across the tree once this is merged 🎉

**Review by paparodeo — APPROVED** [review]

**Comment by emilazy** [comment]
I had a whole thing written up but my browser ate it, so, in short:

* `{pkgsLLVM.hello,firefox,firefox-beta,firefox-devedition,firefox-esr-128,thunderbird,thunderbird-128,librewolf,floorp}` build successfully on `aarch64-linux`, covering Rust, WASI, Mesa, and consequently most of the LLVM surface of a typical Linux desktop.
* Successfully bootstraps on `x86_64-darwin`.
* Everything we’ve fixed here builds on `aarch64-darwin`, and I also built `{nodejs_22,openpam,emacs{28,29},fontforge,swift,man-db,boost,abseil-cpp}` to check that we didn’t need more fixes from Randy’s old branch. @paparodeo has built a ton of stuff on `aarch64-darwin` including OCaml ecosystem things.

There’s bound to be regressions, but I think this is as ready as it’ll ever be. I wish I understood why we need the Firefox workaround, but it works. I believe the only known regression at this point is `nodejs_{18,20}` https://github.com/NixOS/nixpkgs/pull/356408 still needs review and it’d be nice to devendor Zlib there, but it doesn’t affect any default Node versions as of this `staging` cycle so it can wait until after this is merged.

It may or may not be easier to review this as the constituent PRs that went into it.

**Comment by mweinelt** [comment]
Thanks for the due diligence on Firefox. It is appreciated!

**Comment by paparodeo** [comment]
<details><summary>FWIW these are the packages I built on aarch64 and x64 darwin</summary>

```nix
{ pkgs ? import ./. {} }:

let
  channel-blockers = with pkgs; [
    cachix
    cargo
    #darwin.linux-builder
    devenv
    freshBootstrapTools.test
    git
    go
    inkscape
    mariadb
    nix
    nix-info
    nix-info-tested
    nixpkgs-review
    pandoc
    python3
    qt5.qtmultimedia
    #qt6.qtmultimedia
    vim
  ];
  gstreamer-env = let
    inherit (pkgs)
      buildEnv
      glib-networking
      gst_all_1
      wrapGAppsNoGuiHook
    ;
  in
  buildEnv rec {
    name = "gstreamer-env";
    paths = with gst_all_1; [
      gst-devtools
      gst-editing-services
      gst-libav
      gst-plugins-bad
      gst-plugins-base
      gst-plugins-good
      #gst-plugins-rs
      gst-plugins-ugly
      #gst-plugins-viperfx
      gstreamer
      #gstreamermm
      #gst-vaapi
    ];
    nativeBuildInputs = [ wrapGAppsNoGuiHook ];
    buildInputs = paths ++ [ glib-networking ];
    extraOutputsToInstall = [ "out" "bin" "man" ];
    pathsToLink = [ "/bin" "/share/man/man1" ];
    postBuild = ''
      rm -- $out/bin/.*
      gappsWrapperArgsHook
      wrapGAppsHook
    '';
  };
in
  pkgs.mkShell {
  packages = channel-blockers ++ (with pkgs; [
    R
    cargo
    cpio
    cscope
    direnv
    duckdb
    fd
    ffmpeg_7
    freshBootstrapTools.test
    fzf
    gimp
    gstreamer-env
    jq
    jujutsu
    kdiff3
    moreutils
    nix
    nix-output-monitor
    ocaml
    octave
    pandoc
    patchutils
    ripgrep
    rustc
    screen
    sqlite
    swift
    tkdiff
    transmission_4-qt
    transmission_4-gtk
    tree
    uv
    vim-full
    zsh
  ]) ++ (with pkgs.ocaml-ng.ocamlPackages; [
    containers
    dune_3
    findlib
    fmt
    menhir
    merlin
    ounit2
    utop
  ]) ++ (with pkgs.python3Packages; [
    clarabel
    cvxpy
    matplotlib
    numpy
    pandas
    polars
    scs
  ]);
}
```
</details>

File: pkgs/development/compilers/gcc/patches/default.nix
Comment (paparodeo) [inline pkgs/development/compilers/gcc/patches/default.nix:166]:
```suggestion
```

there shouldn't be a `.cfi_startproc-reorderxxxx` patch for gcc-14. it is should be part of Iains patch

Comment (emilazy) [reply pkgs/development/compilers/gcc/patches/default.nix]: Fixed.

**Comment by alyssais** [comment]
Ack.

**Comment by trofi** [comment]
Bisect says that da65c302a18a59c48111bdf219de09cd4b314ce9 `openjpeg: fix up tests` broke test in `staging` on `x86_64-linux` as:

```
$ nix build --no-link -f. -L pkgsi686Linux.openjpeg
...
openjpeg> 99% tests passed, 3 tests failed out of 1528
openjpeg> Total Test time (real) =  13.90 sec
openjpeg> The following tests FAILED:
openjpeg>       104 - NR-JP2-file3.jp2-compare2base (Failed)
openjpeg>       1188 - NR-DEC-issue414.jp2-110-decode-md5 (Failed)
openjpeg>       1227 - NR-DEC-p1_04.j2k-134-decode-md5 (Failed)
```

**Pull request merged**
